_modules/ext_module_qvm.py

```diff
@@ -941,8 +941,203 @@ def prefs(vmname, *varargs, **kwargs):
+    kwargs.setdefault('status-mode', 'all')
+
+    qvm = _QVMBase('qvm.devices', **kwargs)
+    qvm.parser.add_argument(
+        'vmname',
+        action=_VMAction,
+        help='Virtual machine name'
+    )
+    qvm.parser.add_argument('--list', nargs='*', help='List devices')
+    qvm.parser.add_argument(
+        '--attach',
+        nargs='*',
+        default=[],
+        help='List of devices to attach'
+    )
+    qvm.parser.add_argument(
+        '--detach',
+        nargs='*',
+        default=[],
+        help='List of devices to detach'
+    )
+
+    args = qvm.parse_args(vmname, *varargs, **kwargs)
+
+    current_devices = []
+    for device_type in ['pci', 'block', 'usb', 'bridge', 'mic']:
+        for device in args.vm.devices[device_type].assignments():
+            current_devices.append(
+                {'device_type': device_type, 'backend': device.backend_domain.name, 'dev_id': device.ident,
+                 'options': device.options})
+
+    # Return all current devices if a 'list' only was selected
+    if args.list is not None or not (args.attach or args.detach):
+        message = []
+        for device in current_devices:
+            msg_options = '(' + ', '.join('{}={}'.format(key, value) for key, value in device['options'].items()) + ')'
+            message.append(
+                '    ' + device['device_type'] + ':' + device['backend'] + ':' + device['dev_id'] + ' ' + msg_options)
+
+        return {'result': True, 'comment': '[ATTACHED]:\n' + '\n'.join(message)}
+
+    # Helping function for parsing devices in args.attach and args.detach
+    def parse_device(raw_dev):
+        """
+        Example:
+            args.attach = [{u'bridge:sys-bridge:bridge0': [{u'ip': u'192.168.0.1'}, {u'netmask': u'255.255.255.0'}]}]
+        """
+        device = None
+
+        inline_dev = list(raw_dev.keys())[0]
+        inline_dev_split = inline_dev.split(':')
+
+        if len(inline_dev_split) == 3:
+            device = {'device_type': inline_dev_split[0], 'backend': inline_dev_split[1],
+                      'dev_id': inline_dev_split[2], 'options': {}}
+            for opt in raw_dev[inline_dev]:
+                device['options'].update(opt)
+        else:
+            raise SaltInvocationError(
+                message="Missing either 'device_type', 'backend' or 'dev_id' in '%s'" % inline_dev
+            )
+
+        return device
+
+    # post-process argparse 'attach' and 'detach' values converted from string
+    if args.attach:
+        args.attach = [parse_device(raw_dev) for raw_dev in json.loads(args.attach[0])]
+
+    if args.detach:
+        args.detach = [parse_device(raw_dev) for raw_dev in json.loads(args.detach[0])]
+
+    for device in args.attach:
+        device_type = device['device_type']
+
+        if device['options'].get('pci_strictreset', False):
+            device['options']['no-strict-reset'] = not device['options']['pci_strictreset']
+            del device['options']['pci_strictreset']
```

Comment:
`pci_strictreset` is a legacy VM option, not device option. It needs to stay in `prefs` when dealing with PCI devices for compatibility reasons, but no need to add it to this new code.